openid-connect-server/src/main/java/org/mitre/openid/connect/assertion/JWTBearerAuthenticationProvider.java

```diff
@@ -91,57 +91,60 @@ public Authentication authenticate(Authentication authentication) throws Authent
 			JWT jwt = jwtAuth.getJwt();
 			JWTClaimsSet jwtClaims = jwt.getJWTClaimsSet();
 
+			if (!(jwt instanceof SignedJWT)) {
+				throw new AuthenticationServiceException("Unsupported JWT type: " + jwt.getClass().getName());
+			}
+
 			// check the signature with nimbus
-			if (jwt instanceof SignedJWT) {
-				SignedJWT jws = (SignedJWT)jwt;
+			SignedJWT jws = (SignedJWT) jwt;
+
+			JWSAlgorithm alg = jws.getHeader().getAlgorithm();
+
+			if (client.getTokenEndpointAuthSigningAlg() != null &&
+					!client.getTokenEndpointAuthSigningAlg().equals(alg)) {
+				throw new AuthenticationServiceException("Client's registered token endpoint signing algorithm (" + client.getTokenEndpointAuthSigningAlg()
+						+ ") does not match token's actual algorithm (" + alg.getName() + ")");
+			}
+
+			if (client.getTokenEndpointAuthMethod() == null ||
+					client.getTokenEndpointAuthMethod().equals(AuthMethod.NONE) ||
+					client.getTokenEndpointAuthMethod().equals(AuthMethod.SECRET_BASIC) ||
+					client.getTokenEndpointAuthMethod().equals(AuthMethod.SECRET_POST)) {
+
+				// this client doesn't support this type of authentication
+				throw new AuthenticationServiceException("Client does not support this authentication method.");
+
+			} else if ((client.getTokenEndpointAuthMethod().equals(AuthMethod.PRIVATE_KEY) &&
+					(alg.equals(JWSAlgorithm.RS256)
+							|| alg.equals(JWSAlgorithm.RS384)
+							|| alg.equals(JWSAlgorithm.RS512)
+							|| alg.equals(JWSAlgorithm.ES256)
+							|| alg.equals(JWSAlgorithm.ES384)
+							|| alg.equals(JWSAlgorithm.ES512)
+							|| alg.equals(JWSAlgorithm.PS256)
+							|| alg.equals(JWSAlgorithm.PS384)
+							|| alg.equals(JWSAlgorithm.PS512)))
+					|| (client.getTokenEndpointAuthMethod().equals(AuthMethod.SECRET_JWT) &&
+					(alg.equals(JWSAlgorithm.HS256)
+							|| alg.equals(JWSAlgorithm.HS384)
+							|| alg.equals(JWSAlgorithm.HS512)))) {
+
+				// double-check the method is asymmetrical if we're in HEART mode
+				if (config.isHeartMode() && !client.getTokenEndpointAuthMethod().equals(AuthMethod.PRIVATE_KEY)) {
+					throw new AuthenticationServiceException("[HEART mode] Invalid authentication method");
+				}
 
-				JWSAlgorithm alg = jws.getHeader().getAlgorithm();
+				JWTSigningAndValidationService validator = validators.getValidator(client, alg);
 
-				if (client.getTokenEndpointAuthSigningAlg() != null &&
-						!client.getTokenEndpointAuthSigningAlg().equals(alg)) {
-					throw new InvalidClientException("Client's registered request object signing algorithm (" + client.getRequestObjectSigningAlg() + ") does not match request object's actual algorithm (" + alg.getName() + ")");
```

Comment:
Exception class and message was incorrect - this was probably a copy-paste from ConnectOAuth2RequestFactory. The problem was that if this was thrown, it was then caught in this method and UsernameNotFoundException was thrown with a message that client couldn't be found.